### PR TITLE
Add query limit as http header

### PR DESCRIPTION
**Pull request by djaney**
#16 Pagination
#17 Ordering

#### Proposed Changes
* Add limit query as http header called Query-Limit and Query-Offset
* Add order by query as http header called Query-Order

### CODE REVIEW

**Comment by TobiasHauck** [comment]
Hi Djane thanks for opening the PR!
You're absolutely right: We are missing pagination information as well as ordering information and the headers are the right place to add them. Maybe we should rename ```Query-Limit``` to ```Limit```, ```Query-Offset``` to ```Offset``` and ```Query-Order``` to ```Order```.

Would you also be so kind to add the missing tests for the new feature? Thank you for contributing!

File: Factory/RequestFactory.php [hunk 1]
```diff
@@ -50,6 +52,12 @@ public function createOne(array $tokens, $apiUrl, array $options) {
         $query   = HttpQuery::create($tokens);
         $payload = $method === HttpMethods::GET || $method === HttpMethods::DELETE ? null : Payload::create($tokens);
 
+        $options['CURLOPT_HTTPHEADER'] = array_merge(
```
Comment (TobiasHauck) [inline Factory/RequestFactory.php]: I don't like the whole concept for handling HTTP headers right now. That's something about my architecture and not about your PR as you did right.

Let's build a new type HttpHeader. Input is the options array. Output is the options array containing just one entry with the key ```CURLOPT_HTTPHEADER```. First of all we move https://github.com/CircleOfNice/DoctrineRestDriver/blob/master/Types/CurlOptions.php lines 68,69 to the HttpHeader type to make sure that the HttpHeaders are in the right format (```String -> NotNil```).
Afterwards you can put the stuff you are actually doing here into the HttpHeader type. The ResultFactory will just call ```$options = array_merge($options, HttpHeader::create($options);)```.

Finally we need to improve the tests and add a new test for the HttpHeader type.

File: Types/LimitHttpHeader.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+use Circle\DoctrineRestDriver\Enums\SqlOperations;
+use Circle\DoctrineRestDriver\Validation\Assertions;
+
+/**
+ * LimitHttpHeader type
+ *
+ * @author    Djane Rey Mabelin <thedjaney@gmail.com>
+ * @copyright 2016
+ */
+class LimitHttpHeader {
```
Comment (TobiasHauck) [inline Types/LimitHttpHeader.php]: Let's call it PaginationHeaders

File: Types/LimitHttpHeader.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+     * clause of the parsed sql tokens
+     *
+     * @param  array $tokens
+     * @return string
+     *
+     * @SuppressWarnings("PHPMD.StaticAccess")
+     */
+    public static function create(array $tokens) {
+        Assertions::assertHashMap('tokens', $tokens);
+        $arr = [];
```
Comment (TobiasHauck) [inline Types/LimitHttpHeader.php]: $headers might be a better name

Comment (TobiasHauck) [inline Types/LimitHttpHeader.php]: ```if (empty($tokens['LIMIT'])) return [];```

File: Types/LimitHttpHeader.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        if(isset($tokens['LIMIT']['rowcount'])){
```
Comment (TobiasHauck) [inline Types/LimitHttpHeader.php]:
```
$headers = [];
array_push($headers, 'Limit: ' . $tokens['LIMIT']['rowcount']);
array_push($headers, 'Offset: ' . $tokens['LIMIT']['offset']);

return $headers;
```

File: Types/OrderHttpHeader.php [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+use Circle\DoctrineRestDriver\Enums\SqlOperations;
+use Circle\DoctrineRestDriver\Validation\Assertions;
+
+/**
+ * OrderHttpHeader type
+ *
+ * @author    Djane Rey Mabelin <thedjaney@gmail.com>
+ * @copyright 2016
+ */
+class OrderHttpHeader {
```
Comment (TobiasHauck) [inline Types/OrderHttpHeader.php]: What about the name "OrderingHeaders" according to "PaginationHeaders"?

File: Types/OrderHttpHeader.php [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+     * clause of the parsed sql tokens
+     *
+     * @param  array $tokens
+     * @return string
+     *
+     * @SuppressWarnings("PHPMD.StaticAccess")
+     */
+    public static function create(array $tokens) {
+        Assertions::assertHashMap('tokens', $tokens);
+        $arr = [];
```
Comment (TobiasHauck) [inline Types/OrderHttpHeader.php]:
```
if (empty($tokens['order'])) return [];
```

File: Types/OrderHttpHeader.php [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        if(isset($tokens['ORDER'])){
+            $orderQueryArr = [];
+            foreach($tokens['ORDER'] as $order){
```
Comment (TobiasHauck) [inline Types/OrderHttpHeader.php]: In this repo we are trying not to use any foreach, so may you be so kind to rewrite it with [ array functions ] (http://php.net/manual/de/ref.array.php)? Thank you :-)

File: Types/OrderHttpHeader.php [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+                $query = null;
+                if( isset($order['no_quotes']['parts'][1]) ){
```
Comment (TobiasHauck) [inline Types/OrderHttpHeader.php]: Just for your notice: Please use if only in combination with return. If you want to know WHY just let me know and I'll give you an understandable explanation.
What you can do is using the ternary operator instead:

```
$query = isset($order['no_quotes']['parts'][1]) ? $order['no_quotes']['parts'][1] : null;
```

File: Types/OrderHttpHeader.php [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+                    $query = $order['no_quotes']['parts'][1];
+                }
+                if($query && isset($order['direction'])){
```
Comment (TobiasHauck) [inline Types/OrderHttpHeader.php]: same here

File: Factory/RequestFactory.php (hunk 1 above)
Comment (djaney) [inline Factory/RequestFactory.php]: Hi! Thanks for reviewing the code.

I'm a bit confused about this line. Or did you mean that the Input is the $tokens?

> Let's build a new type HttpHeader. Input is the options array. Output is the options array containing just one entry with the key CURLOPT_HTTPHEADER.

Comment (TobiasHauck) [inline Factory/RequestFactory.php]: Oh yes sorry it was the tokens array I mentioned and not the options array. Was my mistake

Comment (TobiasHauck) [inline Factory/RequestFactory.php]: Ok now I am confused, too.

Let's make it clear:
HttpHeader.create:
(array $options, array $tokens) -> Options

```typedef Options {
    CURLOPT_HTTPHEADER: array
}
```

Comment (djaney) [inline Factory/RequestFactory.php]: Yes. That's clear. Thanks.

File: Tests/Types/HttpHeaderTest.php [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    private $expected;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setUp() {
+        $this->options = [
+            'security_strategy'  => 'none',
+            'CURLOPT_MAXREDIRS'  => 22,
+            'CURLOPT_HTTPHEADER' => ['Content-Type: text/plain']
```
Comment (TobiasHauck) [inline Tests/Types/HttpHeaderTest.php]: The input type for CURLOPT_HTTPHEADER is String and not List at this point.
The problem is caused by the Doctrine config settings. All driver options must be scalars, so we have to use a seperated string. So when creating the HTTP Headers we have to transform the string value into a list. I'm not sure about the seperation character, so let's discuss it. What about ;; ?

File: Types/HttpHeader.php [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace Circle\DoctrineRestDriver\Types;
+
+/**
+ * Adds query related headers to the CURLOPT_HTTPHEADER
+ *
+ * @author    Djane Rey Mabelin <thedjaney@gmail.com>
+ * @copyright 2016
+ */
+class HttpHeader{
```
Comment (TobiasHauck) [inline Types/HttpHeader.php]: ``` class HttpHeader { ```

File: Types/HttpHeader.php [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+     * Returns an array containing CURLOPT_HTTPHEADER options that can be added to
+     * the PHP internal curl library by using curl_setopt_array
+     *
+     * @param  array $options
+     * @return array
+     *
+     * @SuppressWarnings("PHPMD.StaticAccess")
+     */
+    public static function create(array $options, array $tokens) {
+
```
Comment (TobiasHauck) [inline Types/HttpHeader.php]: Remove the empty line

File: Types/HttpHeader.php [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+        $headers = empty($headers) ? [] : $headers;
+        $headers = is_string($headers) ? explode(',', $headers) : $headers;
+
+        $headers = array_merge(
+            $headers,
+            PaginationHeaders::create($tokens),
+            OrderingHeaders::create($tokens)
+        );
+        return [
+            'CURLOPT_HTTPHEADER'=>$headers
```
Comment (TobiasHauck) [inline Types/HttpHeader.php]: ```'CURLOPT_HTTPHEADER' => $headers```

File: Types/OrderingHeaders.php [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+     *
+     * @SuppressWarnings("PHPMD.StaticAccess")
+     */
+    public static function create(array $tokens) {
+        Assertions::assertHashMap('tokens', $tokens);
+
+        if (empty($tokens['ORDER'])) return [];
+
+        $headers = [];
+        if(isset($tokens['ORDER'])){
```
Comment (TobiasHauck) [inline Types/OrderingHeaders.php]: The if is not needed any more thanks to ```if (empty($tokens['ORDER'])) return [];```

File: Types/PaginationHeaders.php [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+     * @SuppressWarnings("PHPMD.StaticAccess")
+     */
+    public static function create(array $tokens) {
+        Assertions::assertHashMap('tokens', $tokens);
+
+        if (empty($tokens['LIMIT'])) return [];
+
+        $headers = [];
+        array_push($headers, 'Limit: ' . $tokens['LIMIT']['rowcount']);
+        if(!empty($tokens['LIMIT']['offset'])){
```
Comment (TobiasHauck) [inline Types/PaginationHeaders.php]:
```
if(empty($tokens['LIMIT']['offset'])) return $headers;
array_push($headers, 'Offset: ' . $tokens['LIMIT']['offset']);
return $headers;
```

File: Tests/Types/HttpHeaderTest.php (hunk 10 above)
Comment (djaney) [inline Tests/Types/HttpHeaderTest.php]: CurlOptionsTest also has this but since we use HttpHeader before passing to CurlOptions, it should just be an array. Does that make sense?

Comment (TobiasHauck) [inline Tests/Types/HttpHeaderTest.php]: Well as you can see the type system of PHP is a bitch :D If we had well-formed types we should not have this discussion.
I'd say that CurlOptions awaits CURLOPT_HTTPHEADER to be an array, so the HttpHeader type must handle the transformation from string to array. So your idea makes sense.

File: Tests/Types/HttpHeaderTest.php [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+
+    /**
+     * @test
+     * @group  unit
+     * @covers ::create
+     *
+     * @SuppressWarnings("PHPMD.StaticAccess")
+     */
+    public function create() {
+        $query    = 'SELECT name FROM products WHERE id=1';
```
Comment (TobiasHauck) [inline Tests/Types/HttpHeaderTest.php]: May you be so kind to align all equal signs? :S Sorry for being a perfectionist, but I truly believe in perfectionism as the only way to get rid of bugs

File: Tests/Types/OrderingHeadersTest.php [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+     * @test
+     * @group  unit
+     * @covers ::create
+     *
+     * @SuppressWarnings("PHPMD.StaticAccess")
+     */
+    public function create() {
+        $query    = 'SELECT name FROM products a ORDER BY name ASC';
+        $parser   = new PHPSQLParser();
+        $token = $parser->parse($query);
```
Comment (TobiasHauck) [inline Tests/Types/OrderingHeadersTest.php]: same here with the equal signs

File: Types/OrderingHeaders.php [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+     */
+    public static function create(array $tokens) {
+        Assertions::assertHashMap('tokens', $tokens);
+
+        if (empty($tokens['ORDER'])) return [];
+        
+        $headers = [];
+        $orderQueryArr = array_map(function($order){
+            $query = null;
+            $field = end($order['no_quotes']['parts']);
```
Comment (TobiasHauck) [inline Types/OrderingHeaders.php]:
```
$query = end($order['no_quotes']['parts']);
if (empty($query)) return null;
return !isset($order['direction']) ? $query : $query . ' ' . $order['direction'];
```

File: Tests/Types/HttpHeaderTest.php (hunk 16 above)
Comment (djaney) [inline Tests/Types/HttpHeaderTest.php]: No problem :D

File: Types/OrderingHeaders.php [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+     * @SuppressWarnings("PHPMD.StaticAccess")
+     */
+    public static function create(array $tokens) {
+        Assertions::assertHashMap('tokens', $tokens);
+
+        if (empty($tokens['ORDER'])) return [];
+
+        $headers = [];
+        $orderQueryArr = array_map(function($order){
+            $query = null;
```
Comment (TobiasHauck) [inline Types/OrderingHeaders.php]: Why is that needed?

File: Tests/Types/HttpHeaderTest.php [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+     */
+    public function setUp() {
+        $this->options = [
+            'security_strategy'  => 'none',
+            'CURLOPT_MAXREDIRS'  => 22,
+            'CURLOPT_HTTPHEADER' => 'Content-Type: text/plain'
+        ];
+
+        $this->expected = [
+            'CURLOPT_HTTPHEADER'=> ['Content-Type: text/plain']
```
Comment (TobiasHauck) [inline Tests/Types/HttpHeaderTest.php]: space before ```=> ```

File: Types/OrderingHeaders.php (hunk 19 above)
Comment (djaney) [inline Types/OrderingHeaders.php]: ah. a residue of the old code. will remove that

File: Tests/Types/HttpHeaderTest.php (hunk 16 above)
Comment (TobiasHauck) [inline Tests/Types/HttpHeaderTest.php]: Ok my fault: I didn't tell you how to align the equal signs. My intention was to have at least one assignment that has exactly one space between the var and the equal. All the other equals should be aligned with that one.

Have a look at https://github.com/CircleOfNice/DoctrineRestDriver/blob/master/Tests/Transformers/MysqlToRequestTest.php lines 136, 137. We are almost done with this PR :D

File: Tests/Types/PaginationHeadersTest.php [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+
+    /**
+     * @test
+     * @group  unit
+     * @covers ::create
+     *
+     * @SuppressWarnings("PHPMD.StaticAccess")
+     */
+    public function create() {
+        $query    = 'SELECT name FROM products LIMIT 10, 10';
```
Comment (TobiasHauck) [inline Tests/Types/PaginationHeadersTest.php]: alignment

File: Types/HttpHeader.php (hunk 11 above)
Comment (TobiasHauck) [inline Types/HttpHeader.php]: Please add the space here - we're almost done :P

**Comment by TobiasHauck** [comment]
Ok if you have changed all my suggestions this one is ready to merge - thanks a lot, you did a great job!

File: Types/OrderingHeaders.php (hunk 19 above)
Comment (TobiasHauck) [inline Types/OrderingHeaders.php]: Ok :D You might have had a reason for this line, who knows? ;)

File: Tests/Types/PaginationHeadersTest.php [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+
+use Circle\DoctrineRestDriver\Types\PaginationHeaders;
+use PHPSQLParser\PHPSQLParser;
+/**
+ * Tests the curl options
+ *
+ * @author    Tobias Hauck <tobias@circle.ai>
+ * @copyright 2015 TeeAge-Beatz UG
+ *
+ * @coversDefaultClass Circle\DoctrineRestDriver\Types\CurlOptions
```
Comment (TobiasHauck) [inline Tests/Types/PaginationHeadersTest.php]:
```
@coversDefaultClass Circle\DoctrineRestDriver\Types\CurlOptions\PaginationHeaders
```

otherwise the code coverage report is buggy :D

File: Tests/Types/PaginationHeadersTest.php [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with DoctrineRestDriver.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace Circle\DoctrineRestDriver\Tests\Types;
+
+use Circle\DoctrineRestDriver\Types\PaginationHeaders;
+use PHPSQLParser\PHPSQLParser;
+/**
+ * Tests the curl options
```
Comment (TobiasHauck) [inline Tests/Types/PaginationHeadersTest.php]: Tests the pagination headers

File: Tests/Types/OrderingHeadersTest.php [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+
+use Circle\DoctrineRestDriver\Types\OrderingHeaders;
+use PHPSQLParser\PHPSQLParser;
+/**
+ * Tests the curl options
+ *
+ * @author    Tobias Hauck <tobias@circle.ai>
+ * @copyright 2015 TeeAge-Beatz UG
+ *
+ * @coversDefaultClass Circle\DoctrineRestDriver\Types\CurlOptions
```
Comment (TobiasHauck) [inline Tests/Types/OrderingHeadersTest.php]:
```
@coversDefaultClass Circle\DoctrineRestDriver\Types\OrderingHeaders
```

File: Tests/Types/OrderingHeadersTest.php [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with DoctrineRestDriver.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace Circle\DoctrineRestDriver\Tests\Types;
+
+use Circle\DoctrineRestDriver\Types\OrderingHeaders;
+use PHPSQLParser\PHPSQLParser;
+/**
+ * Tests the curl options
```
Comment (TobiasHauck) [inline Tests/Types/OrderingHeadersTest.php]: Tests the ordering headers

File: Tests/Types/HttpHeaderTest.php [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+
+use Circle\DoctrineRestDriver\Types\HttpHeader;
+use PHPSQLParser\PHPSQLParser;
+/**
+ * Tests the curl options
+ *
+ * @author    Tobias Hauck <tobias@circle.ai>
+ * @copyright 2015 TeeAge-Beatz UG
+ *
+ * @coversDefaultClass Circle\DoctrineRestDriver\Types\CurlOptions
```
Comment (TobiasHauck) [inline Tests/Types/HttpHeaderTest.php]: ```@coversDefaultClass Circle\DoctrineRestDriver\Types\HttpHeaders```

File: Tests/Types/HttpHeaderTest.php [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with DoctrineRestDriver.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace Circle\DoctrineRestDriver\Tests\Types;
+
+use Circle\DoctrineRestDriver\Types\HttpHeader;
+use PHPSQLParser\PHPSQLParser;
+/**
+ * Tests the curl options
```
Comment (TobiasHauck) [inline Tests/Types/HttpHeaderTest.php]: Tests the http header

**Comment by TobiasHauck** [comment]
Just checked the code coverage to merge the PR and realized it was not 100% ... 😢 This is when I found these last errors :D

**Comment by TobiasHauck** [comment]
Well if you like to you choose the next release title. It has to be related to the TV series house of cards.

**Comment by djaney** [comment]
I added the coverage annotations. My coverage is not working so I don't know if there are more

**Comment by TobiasHauck** [comment]
That's all. Coverage is at 100% again. So please update the comments and we are ready. What about the release name?

**Comment by djaney** [comment]
I'm not really on the house of cards. Does "The Washington Herald" sound ok?

**Comment by TobiasHauck** [comment]
yes great, let's call it "The Washington Herald"
